packages/teleterm/src/mainProcess/runtimeSettings.ts

```diff
@@ -101,18 +104,73 @@ function getAssetPath(...paths: string[]): string {
 
 function getDefaultShell(): string {
   const logger = new Logger();
-  const fallbackShell = 'bash';
-  const { shell } = os.userInfo();
+  switch (process.platform) {
+    case 'linux':
+    case 'darwin': {
+      const fallbackShell = 'bash';
+      const { shell } = os.userInfo();
+
+      if (!shell) {
+        logger.error(
+          `Failed to read ${process.platform} platform default shell, using fallback: ${fallbackShell}.\n`
+        );
+
+        return fallbackShell;
+      }
+
+      return shell;
+    }
+    case 'win32':
+      return 'powershell.exe';
+  }
+}
 
-  if (!shell) {
-    logger.error(
-      `Failed to read ${process.platform} platform default shell, using fallback: ${fallbackShell}.\n`
-    );
+async function getGrpcAdresses(): Promise<{
+  tshAddress: string;
+  sharedAddress: string;
+}> {
+  switch (process.platform) {
+    case 'win32': {
+      const [tshPort, sharedPort] = await Promise.all([
+        getAvailablePort(),
+        getAvailablePort(),
+      ]);
+      return {
+        tshAddress: `localhost:${tshPort}`,
+        sharedAddress: `localhost:${sharedPort}`,
+      };
+    }
+    case 'linux':
+    case 'darwin':
+      return {
+        tshAddress: getUnixSocketNetworkAddress('tsh.socket'),
+        sharedAddress: getUnixSocketNetworkAddress('shared.socket'),
+      };
+  }
+}
+
+function getUnixSocketNetworkAddress(socketName: string) {
+  const unixSocketPath = path.resolve(app.getPath('userData'), socketName);
 
-    return fallbackShell;
+  // try to cleanup after previous process that unexpectedly crashed
+  if (fs.existsSync(unixSocketPath)) {
+    fs.unlinkSync(unixSocketPath);
   }
 
-  return shell;
+  return `unix://${path.resolve(app.getPath('userData'), socketName)}`;
+}
+
+async function getAvailablePort(): Promise<number> {
+  const server = net.createServer();
+  return new Promise(resolve => {
+    // OS will find a free port automatically
+    server.listen({ port: 0 }, () => {
+      const { port } = server.address() as net.AddressInfo;
+      server.close(() => {
+        resolve(port);
+      });
+    });
```

Comment:
I saw a couple of answers on SO that suggested using a command for that on Windows (`netstat` I think?). The downside is that the output might be different depending on the system language of the user. A Powershell command has also been mentioned which might be more reliable.

Reply:
I considered doing this like you wrote, but unfortunately it will increase the complexity of establishing the connections significantly. 
1. Currently, both the UDS and TCP connections are created in the same way, so both clients just read addresses from `runtimeSettings` and connect to them. After changing this, we would have to use different logic for both types of connections.
2. Checking the port by PID requires spawning a new process (and probably it is not a solution, I guess that `netstat` returns all ports that a process use?)
3. The option that would work is to return ports from processes via stdout, but it requires looking for some string in process output and blocking the creation of clients until we get the correct addresses. 

I was curious how the OS assigns port numbers, so I performed quick test. I invoked `getAvailablePort()` and then wait 300 ms, 10 times. The ports I got were not repeated, OS assigned a new port for each connection, event the previous one was already closed. 
This doesn't give 100% confidence, for sure some apps do not relay on OS, instead they just randomly create the port number.

It would be really good to have this done right, but it requires quite a few changes and I feel like this problem is unlikely to happen... I don't know, @zmb3 do you have any opinion?

Reply:
I said this in DMs between us but I'd rather do it properly rather than keep our fingers crossed each time someone launches the app that everything goes okay. ;)

I'm not sure though what's the best way for a program to advertise what port it occupies. Using a system like Bonjour is probably an overkill in this situation. Reading from stdout of the process seems alright but idk if there's a better way to do it. Those processes will always be launched by our app only and never by the user, so we can even emit the port as a specific string like `CONNECT_GRPC_PORT: 1337`.

---

When it comes to changes necessary in our startup logic: the network addresses are supplied in the runtime settings right now and the runtime settings is the first thing we load when we start the app. Changing this logic would be in fact quite cumbersome.

However, what if instead we just changed the logic that sets up the clients for tshd and shared process? `MainProcess.create` would need to become an async function that returns the ports used by gRPC servers. Then we'd need to change only the code which starts the clients rather than changing all code which reads the runtime settings.

Reply:
Ok, I'm thinking about this again now, and it seems that we can do it in a way where there is no difference between UDS and TCP. Processes can return whole addresses like `tcp://host:port` or `unix://path` which we will provide to clients. The idea with changing `MainProcess.create` to be async makes sense, I'll work on that. Thanks for your comments!

Reply:
In the end, I decided to return only the port from the child processes. Returning the full address was a bit problematic, for example, tshd retuned `127.0.0.1` instead of `localhost` (`grpc-js` uses DNS to resolve the address) and I had to remove `tcp://` from returned address manually. 

I modified @ravicious idea a bit. Because we have to provide the resolved addresses to the `preload` script, we need to add a new `ipcMain` handler. That handler can be the only place that interacts with the logic that resolves the addresses. By doing this, we don't have to change the code in `main.ts`.

Reply:
This sounds like a good approach. I agree with Rafal that we should do it right rather than hope that the ports are still free by the time we launch tshd.